pkg/sql/lease.go

```diff
@@ -1115,16 +1141,35 @@ func nameMatchesTable(table *sqlbase.TableDescriptor, dbID sqlbase.ID, tableName
 // the timestamp. It returns the table descriptor and a expiration time.
 // A transaction using this descriptor must ensure that its
 // commit-timestamp < expiration-time. Care must be taken to not modify
-// the returned descriptor.
+// the returned descriptor. AcquireByName will renew a lease that is about to
+// expire to prevent blocking when no lease is available.
 func (m *LeaseManager) AcquireByName(
 	ctx context.Context, timestamp hlc.Timestamp, dbID sqlbase.ID, tableName string,
 ) (*sqlbase.TableDescriptor, hlc.Timestamp, error) {
 	// Check if we have cached an ID for this name.
 	tableVersion := m.tableNames.get(dbID, tableName, timestamp)
 	if tableVersion != nil {
 		if !timestamp.Less(tableVersion.ModificationTime) {
+			// Check if we should preemptively acquiring a new lease if this one will
+			// expire soon.
+			if tableVersion.shouldPreemptivelyRenewLease() {
+				t := m.findTableState(tableVersion.ID, true)
+				// innerCtx, span := tracing.ForkCtxSpan(ctx, "[async] lease acquisition")
+				t.mu.group.DoChan("acquire", func() (interface{}, error) {
+					// defer tracing.FinishSpan(span)
+					log.Warningf(context.Background(), "TableDescriptor async renewal")
+					err := t.acquireNodeLease(context.Background(), m, hlc.Timestamp{})
+					if err != nil {
+						log.Error(context.Background(), errors.Wrapf(err, "TableDescriptor async lease acquisition failed"))
+					}
+					return nil, err
+				})
+			}
```

Comment:
I'm not happy about the code above because it is likely to be non performant. Instead I think what we can do is add a timer to tableVersionState and modify m.tablename.get() above to check if the timer is set. If it is not set, it can create a timer using ``` func AfterFunc(d Duration, f func()) *Timer``` from the time package. The function can renew the lease when it fires and reset the timer to nil.

Additionally when the tableVersionState is released the timer should be stopped. If AfterFunc() is called during the stop then it should wait until Afterfunc sets the timer to nil rather than setting it to nil on its own.

Thoughts?

Reply:
I think the timer check and creation don't belong in `m.tableNames.get` because the method is currently is just focused on retrieving the lease from the table cache. Otherwise, a few additional parameters need to get passed in _just_ to do the timer initialization to renew (LeaseManager, table ID).

Instead, the code added there could be replaced with
```
if ...timer == nil {
   timer = timer.AfterFunc(s.durationUntilRenewal(), func() { m.Acquire(ctx, hcl.Timestamp{}, tableVersion.ID) })
}
```

The uncertainty I have is where the timer should live. You suggest that we reset the timer when we release a `tableVersionState`, so in that case, it lives there. If we get a new lease, we will have a new tableVersionState, so won't we end up initializing a second timer? 

(unrelated to your comment) I realized this code could instead be restructured to just call `LeaseManager.Acquire` and instead put the `DoChan`  in `tableState.acquire`. The only addition to `AcquireByName` is to check for preemptive renewal and then call `LeaseManager.Acquire`. That's much cleaner because 1) it's not exposing the `DoChan` internals here, and 2) `LeaseManager.Acquire` already does the exact same table name lookup. I was also confused when `tableState.acquire` wasn't always called previously, so this would make more sense.

That's why I suggested the AfterFunc to be much simpler above.